### ISSUE TITLE
MicronautAwsProxyExceptionHandler giving 500 in case of JSON Mapping exception

### ISSUE THREAD

**Issue by ShashankMishra** [micronaut-projects/micronaut-aws#52]
We are deserialising request json dates with our custom deserialiser.
In case of invalid date format we are throwing our custom exception which ultimately gets wrapped in JSONMappingException and later on to InvalidRequestEventException.

Now since I want to send 400 BAD REQUEST in such cases but **MicronautAwsProxyExceptionHandler** is handling these exception by default and send 

> 500 Internal Server Error 

in place of 

> 400 BAD REQUEST.

We tried creating our own error handler and tried replacing **MicronautAwsProxyExceptionHandler** but that too is not working because this exception handler is directly injected in **MicronautLambdaContainerHandler**.

Flow :
1.  Request forwarded to lambda handler  from our function -

```java
    @Override
    public void handleRequest(InputStream input, OutputStream output, Context context) throws IOException {
        logger.info("Input: {}", input);
        handler.proxyStream(input, output, context);
    } 
```

2.  Gets exception generated due to json deserialisation of wrong format  of request at - **MicronautLambdaContainerHandler.java**

```java
    public void proxyStream(InputStream input, OutputStream output, Context context)
            throws IOException {

        try {
            RequestType request = readerFor(requestTypeClass).readValue(input);
            ResponseType resp = proxy(request, context);

            writerFor(responseTypeClass).writeValue(output, resp);
        } catch (JsonParseException e) {
            log.error("Error while parsing request object stream", e);
            objectMapper().writeValue(output, exceptionHandler.handle(e));
        } catch (JsonMappingException e) {
            log.error("Error while mapping object to RequestType class", e);
            objectMapper().writeValue(output, exceptionHandler.handle(e));
        } finally {
            output.flush();
            output.close();
        }
    }
```
3. MicronautAwsProxyExceptionHandler  handling exception and sending 500 in place of 400 Bad request.

```java
    @Override
    public AwsProxyResponse handle(Throwable ex) {
        LOG.error("Called exception handler for:", ex);

        // adding a print stack trace in case we have no appender or we are running inside SAM local, where need the
        // output to go to the stderr.
        ex.printStackTrace();
        **if (ex instanceof InvalidRequestEventException) {
            return new AwsProxyResponse(500, headers,** getErrorJson(INTERNAL_SERVER_ERROR));
        } else {
            return new AwsProxyResponse(502, headers, getErrorJson(GATEWAY_TIMEOUT_ERROR));
        }
    }
```

**Why is it doing this part, sending 500 for InvalidRequestEventException :** 

> **if (ex instanceof InvalidRequestEventException) {
>             return new AwsProxyResponse(500, headers,**

**Comment by ShashankMishra** [comment]
Any update please?

**Comment by graemerocher** [comment]
The implementation is based on the existing AWS API proxy implementations which do that, but you are welcome to send a PR to make this behaviour customizable

**Comment by pulsarix** [comment]
@ShashankMishra  Did you try to add micronaut-hibernate-validator ,it will return you 400 instead of 500.

**Comment by ShashankMishra** [comment]
> @ShashankMishra Did you try to add `micronaut-hibernate-validator` ,it will return you 400 instead of 500.

Yes but that works when your json payload translated to Java Object, In case of having different Enum values, error comes even before this translation.

**Comment by alvarosanchez** [comment]
@graemerocher can this not be solved using `@Error` handlers in the controller?

**Comment by graemerocher** [comment]
Seems like it should be handled that way but it is not, which is the bug. In the normal HTTP server we have `JsonExceptionHandler` that does this

**Comment by magJ** [comment]
Looking at the PR, it doesn't seem to solve the problem, I believe this issue should be reopened.  

The PR makes changes to the exception handling behaviour within `MicronautLambdaContainerHandler#handleRequest`, but the `InvalidRequestEventException` is thrown before ever entering that method.  

It can only be handled by `MicronautAwsProxyExceptionHandler`, which there appears to be no user way to override.

For reference here is a partial stack trace of a bad request example:
``` 
Caused by: com.fasterxml.jackson.databind.exc.MismatchedInputException: Cannot deserialize instance of `java.lang.Integer` out of VALUE_TRUE token
 at redacted
 at com.fasterxml.jackson.databind.exc.MismatchedInputException.from(MismatchedInputException.java:59) ~[jackson-databind-2.10.3.jar:2.10.3]
 at com.fasterxml.jackson.databind.DeserializationContext.reportInputMismatch(DeserializationContext.java:1445) ~[jackson-databind-2.10.3.jar:2.10.3]
 at com.fasterxml.jackson.databind.DeserializationContext.handleUnexpectedToken(DeserializationContext.java:1219) ~[jackson-databind-2.10.3.jar:2.10.3]
 at com.fasterxml.jackson.databind.DeserializationContext.handleUnexpectedToken(DeserializationContext.java:1129) ~[jackson-databind-2.10.3.jar:2.10.3]
 at com.fasterxml.jackson.databind.deser.std.NumberDeserializers$IntegerDeserializer._parseInteger(NumberDeserializers.java:531) ~[jackson-databind-2.10.3.jar:2.10.3]
 at com.fasterxml.jackson.databind.deser.std.NumberDeserializers$IntegerDeserializer.deserialize(NumberDeserializers.java:474) ~[jackson-databind-2.10.3.jar:2.10.3]
 at com.fasterxml.jackson.databind.deser.std.NumberDeserializers$IntegerDeserializer.deserialize(NumberDeserializers.java:452) ~[jackson-databind-2.10.3.jar:2.10.3]
 at com.fasterxml.jackson.databind.deser.SettableBeanProperty.deserialize(SettableBeanProperty.java:530) ~[jackson-databind-2.10.3.jar:2.10.3]
 at com.fasterxml.jackson.databind.deser.BeanDeserializer._deserializeWithErrorWrapping(BeanDeserializer.java:528) ~[jackson-databind-2.10.3.jar:2.10.3]
 at com.fasterxml.jackson.databind.deser.BeanDeserializer._deserializeUsingPropertyBased(BeanDeserializer.java:417) ~[jackson-databind-2.10.3.jar:2.10.3]
 at com.fasterxml.jackson.databind.deser.BeanDeserializerBase.deserializeFromObjectUsingNonDefault(BeanDeserializerBase.java:1287) ~[jackson-databind-2.10.3.jar:2.10.3]
 at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserializeFromObject(BeanDeserializer.java:326) ~[jackson-databind-2.10.3.jar:2.10.3]
 at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserialize(BeanDeserializer.java:159) ~[jackson-databind-2.10.3.jar:2.10.3]
 at com.fasterxml.jackson.databind.deser.SettableBeanProperty.deserialize(SettableBeanProperty.java:530) ~[jackson-databind-2.10.3.jar:2.10.3]
 at com.fasterxml.jackson.databind.deser.BeanDeserializer._deserializeWithErrorWrapping(BeanDeserializer.java:528) ~[jackson-databind-2.10.3.jar:2.10.3]
 at com.fasterxml.jackson.databind.deser.BeanDeserializer._deserializeUsingPropertyBased(BeanDeserializer.java:417) ~[jackson-databind-2.10.3.jar:2.10.3]
 at com.fasterxml.jackson.databind.deser.BeanDeserializerBase.deserializeFromObjectUsingNonDefault(BeanDeserializerBase.java:1287) ~[jackson-databind-2.10.3.jar:2.10.3]
 at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserializeFromObject(BeanDeserializer.java:326) ~[jackson-databind-2.10.3.jar:2.10.3]
 at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserialize(BeanDeserializer.java:159) ~[jackson-databind-2.10.3.jar:2.10.3]
 at com.fasterxml.jackson.databind.ObjectMapper._readMapAndClose(ObjectMapper.java:4218) ~[jackson-databind-2.10.3.jar:2.10.3]
 at com.fasterxml.jackson.databind.ObjectMapper.readValue(ObjectMapper.java:3214) ~[jackson-databind-2.10.3.jar:2.10.3]
 at com.fasterxml.jackson.databind.ObjectMapper.readValue(ObjectMapper.java:3182) ~[jackson-databind-2.10.3.jar:2.10.3]
 at io.micronaut.jackson.codec.JacksonMediaTypeCodec.decode(JacksonMediaTypeCodec.java:190) ~[micronaut-runtime-1.3.4.jar:1.3.4]
 at io.micronaut.function.aws.proxy.MicronautRequestReader.readRequest(MicronautRequestReader.java:108) ~[micronaut-function-aws-api-proxy-1.3.9.jar:1.3.9]
 ... 13 more
 2020-05-08 04:00:46  ERROR AwsProxyExceptionHandler:73 - Called exception handler for:
 com.amazonaws.serverless.exceptions.InvalidRequestEventException: Invalid Request: Error decoding JSON stream for type [redacted]: Cannot deserialize instance of `java.lang.Integer` out of VALUE_TRUE token
 at redacted
 at io.micronaut.function.aws.proxy.MicronautRequestReader.readRequest(MicronautRequestReader.java:123) ~[micronaut-function-aws-api-proxy-1.3.9.jar:1.3.9]
 at io.micronaut.function.aws.proxy.MicronautRequestReader.readRequest(MicronautRequestReader.java:48) ~[micronaut-function-aws-api-proxy-1.3.9.jar:1.3.9]
 at io.micronaut.function.aws.proxy.AbstractLambdaContainerHandler.proxy(AbstractLambdaContainerHandler.java:204) [micronaut-function-aws-api-proxy-1.3.9.jar:1.3.9]
 at io.micronaut.function.aws.proxy.AbstractLambdaContainerHandler.proxyStream(AbstractLambdaContainerHandler.java:239) [micronaut-function-aws-api-proxy-1.3.9.jar:1.3.9]
```

**Comment by magJ** [comment]
So I have raised the PR #306 to address this, it allows the user to define an exception mapper for decoding errors.  
Ideally these errors would be handled automatically, the same way the netty implementation does, but I don't think I'm the right person to make that change.

**Comment by KnutKnutson** [comment]
What's the release plan for magJ's fix? The PR isn't associated with a milestone.

**Comment by ilopmar** [comment]
The PR was included in `2.0.0.RC1` release: https://github.com/micronaut-projects/micronaut-aws/releases/tag/v2.0.0.RC1

![image](https://user-images.githubusercontent.com/559192/89193873-e61a6b80-d5a6-11ea-8851-511bcf50ebf9.png)
